### DISCUSSION TITLE
How to configure <Image /> to support hiDpi (Retina) images?

### DISCUSSION THREAD

**Discussion by predaytor** [vercel/next.js#28785] [Help] [answered]
I can define image sizes in next.config.js for different viewports.
```
module.exports = {
	images: {
		deviceSizes: [576, 768, 1024, 1440, 1920, 2560, 3840],
	},
};
```
But how, for example, to include a **2x** size image? Do I have to specify an additional size **2x** for each size **1x**? Like:

deviceSizes: [576, **1152**, 768, **1536**, 1024, **2048**, 1440, **2880**, 1920, **3840**, 2560, **5120**, **7680**]

Then, how can I target these sizes, like `srcset` in standart `<img />`?
Thx.

**Comment by predaytor** [comment, marked as answer]
Yeah, just use this way:
```
deviceSizes: [576, 768, 1024, 1440, 1920].flatMap(size => [size, size * 2]),
```

**Reply by icyJoseph** [reply]
Alternatively if `flatMap` is not supported on your node version:

```js
[576, 768, 1024, 1440, 1920, 2560, 3840].reduce((prev, curr) => [...prev, curr, curr * 2], []);
```

Assuming order doesn't matter. It is a pretty small list so spreading prev shouldn't be an issue, otherwise use 

```js
[576, 768, 1024, 1440, 1920, 2560, 3840].reduce((prev, curr) => { prev.push(curr,  curr * 2); return prev; }, []);
```

**Reply by Kyoss79** [reply]
There's screens with pixelratios of 3 nowadays. This method is verbose and ... doesnt really solve the issue.

**Reply by predaytor** [reply]
no need to include, as there will be no difference in image quality between x2 and x3. Apple itself only uses x1, x2 descriptors on its website.

**Reply by seitbekir** [reply]
As I tested on Pixel XL, it works. I use actually following code to fully support devices the way it will looks better.

```html
<source
    media="(resolution > 3x)"
    srcSet="images/LoginPage-logo@4x.png"
/>
```
